aten/src/ATen/core/dispatch/CppSignature.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <typeindex>
+#include <c10/macros/Macros.h>
+#include <c10/util/Metaprogramming.h>
+#include <c10/util/Type.h>
+
+namespace c10 {
+namespace impl {
+
+// A CppSignature object holds RTTI information about a C++ function signature at runtime
+// and can compare them or get a debug-printable name.
+class CAFFE2_API CppSignature final {
+public:
+    CppSignature(const CppSignature&) = default;
+    CppSignature(CppSignature&&) noexcept = default;
+    CppSignature& operator=(const CppSignature&) = default;
+    CppSignature& operator=(CppSignature&&) noexcept = default;
+
+    template<class FuncType>
+    static CppSignature make() {
+        // Normalize functors, lambdas, function pointers, etc. into the plain function type
+        using decayed_function_type = typename guts::infer_function_traits_t<std::decay_t<FuncType>>::func_type;
+
+        return CppSignature(std::type_index(typeid(decayed_function_type)));
```

Comment:
Use of typeid here means that you must compile with RTTI for this to work. We have some settings where we don't compile with RTTI. What happens in those cases?

Reply:
we don't have non-RTTI builds atm